xvec/index.py

```diff
@@ -0,0 +1,193 @@
+from __future__ import annotations
+
+from typing import Any, Hashable, Iterable, Mapping, Sequence
+
+import numpy as np
+import pandas as pd
+import shapely
+from pyproj import CRS
+from xarray import DataArray, Variable
+from xarray.core.indexing import IndexSelResult
+from xarray.indexes import Index, PandasIndex
+
+
+class GeoVectorIndex(Index):
+    """An CRS-aware, Xarray-compatible index for vector geometries.
+
+    This index can be set from any 1-dimensional coordinate of
+    (shapely 2.0) :class:`shapely.Geometry` elements.
+
+    It provides all the basic functionality of an
+    :class:`xarray.indexes.PandasIndex`. In addition, it allows spatial
+    filtering based on geometries (powered by :class:`shapely.STRtree`).
+
+    Parameters
+    ----------
+    index : :class:`xarray.indexes.PandasIndex`
+        An Xarray (pandas) index built from an array-like of
+        :class:`shapely.Geometry` objects.
+    crs : object
+        The coordinate reference system. Any value accepted by
+        :meth:`pyproj.crs.CRS.from_user_input`.
+
+    """
+
+    _index: PandasIndex
+    _sindex: shapely.STRtree | None
+    _crs: CRS
+
+    def __init__(self, index: PandasIndex, crs: CRS):
+        if not np.all(shapely.is_geometry(index.index)):
+            raise ValueError("array must contain shapely.Geometry objects")
+
+        self._crs = CRS.from_user_input(crs)
+        self._index = index
+        self._sindex = None
+
+    @property
+    def crs(self) -> CRS:
+        """Returns the coordinate reference system of the index as a
+        :class:`pyproj.crs.CRS` object.
+        """
+        return self._crs
+
+    @property
+    def sindex(self) -> shapely.STRtree:
+        """Returns the spatial index, i.e., a :class:`shapely.STRtree` object.
+
+        It may build the index before returning it.
+        """
+        if self._sindex is None:
+            self._sindex = shapely.STRtree(self._index.index)
+        return self._sindex
+
+    @classmethod
+    def from_variables(
+        cls,
+        variables: Mapping[Any, Variable],
+        *,
+        options: Mapping[str, Any],
+    ):
+        # TODO: try getting CRS from coordinate attrs or GeometryArray
```

Comment:
```suggestion
        # TODO: try getting CRS from coordinate attrs or GeometryArray or SRID
```

Just so we don't forget about [this](https://shapely.readthedocs.io/en/latest/reference/shapely.get_srid.html?highlight=get_srid#shapely.get_srid) option as well. It is not common but when loading geoms from PostGIS, this may be useful.

Reply:
I wonder if we can somehow get `xarray` to extract the CRS from the `GeometryArray` and store it on the object? That would probably have to be done in a hook that would be defined here... some sort of `array type → hook callable` mapping where the hooks return either the full `Variable` object or a `data, attrs` tuple, maybe?

Also, that would require some thought into how to store the `crs`. Not sure if we could copy whatever `rioxarray` / `odc-geo` are doing, since those seem to have one global `crs` per `DataArray` / `Dataset`?

Reply:
Yes it would be nice to extract the CRS from the GeometryArray, but for dimension coordinates Xarray currently converts the GeometryArray into a plain numpy array, therefore losing the CRS info. We need to fix that when refactoring IndexVariable in Xarray.

> Also, that would require some thought into how to store the crs. Not sure if we could copy whatever rioxarray / odc-geo are doing, since those seem to have one global crs per DataArray / Dataset?

With a `CRSIndex` (https://github.com/corteva/rioxarray/issues/588) or `RasterIndex` built from three coordinates (2D spatial + spatial_ref), having one global crs per DataArray / Dataset rioxarray may not be required anymore and the conventions regarding the coordinate names may even be relaxed.

While for rasters it makes sense to have a scalar coordinate holding the CRS info to avoid duplicating it, for vector data the coordinate <-> index relationship is 1:1 so we could add the CRS info directly as attribute(s) of the vector coordinate.

It may be a good idea to add CRS info (a serializable version) as a coordinate attribute, if not present. So basic CRS info (identifier) would be accessible from the DataArray / Dataset repr, and the full CRS object would be accessible from the index object. E.g.,

```python
>>> ds.geom_coord.attrs["crs"]
'EPSG:4267'
>>> ds.xindexes["geom_coord"].crs
<pyproj.crs.CRS>
```